protocol/streaming/full_node_streaming_manager.go

```diff
@@ -237,74 +277,109 @@ func (sm *FullNodeStreamingManagerImpl) Stop() {
+		{
+			UpdateMessage: &clobtypes.StreamUpdate_OrderbookUpdate{
+				OrderbookUpdate: &clobtypes.StreamOrderbookUpdate{
+					Updates:  v1updates,
+					Snapshot: true,
 				},
-				BlockHeight: blockHeight,
-				ExecMode:    uint32(execMode),
 			},
-		}
-		metrics.IncrCounter(
-			metrics.GrpcAddToSubscriptionChannelCount,
-			1,
+			BlockHeight: blockHeight,
+			ExecMode:    uint32(execMode),
+		},
+	}
+}
+
+func (sm *FullNodeStreamingManagerImpl) getSubaccountStreamUpdates(
+	saUpdates []*satypes.StreamSubaccountUpdate,
+	blockHeight uint32,
+	execMode sdk.ExecMode,
+) []clobtypes.StreamUpdate {
+	streamUpdates := make([]clobtypes.StreamUpdate, 0)
+	for _, saUpdate := range saUpdates {
+		streamUpdates = append(streamUpdates, clobtypes.StreamUpdate{
+			UpdateMessage: &clobtypes.StreamUpdate_SubaccountUpdate{
+				SubaccountUpdate: saUpdate,
+			},
+			BlockHeight: blockHeight,
+			ExecMode:    uint32(execMode),
+		})
+	}
+	return streamUpdates
+}
+
+func (sm *FullNodeStreamingManagerImpl) sendStreamUpdates(
+	subscriptionId uint32,
+	streamUpdates []clobtypes.StreamUpdate,
+) {
+	removeSubscription := false
+	subscription, ok := sm.orderbookSubscriptions[subscriptionId]
+	if !ok {
+		sm.logger.Error(
+			fmt.Sprintf(
+				"Streaming subscription id %+v not found. This should not happen.",
+				subscriptionId,
+			),
 		)
-		select {
-		case subscription.updatesChannel <- streamUpdates:
-		default:
-			sm.logger.Error(
-				fmt.Sprintf(
-					"Streaming subscription id %+v channel full capacity. Dropping subscription connection.",
-					subscriptionId,
-				),
-			)
-			removeSubscription = true
-		}
+		return
+	}
+
+	select {
+	case subscription.updatesChannel <- streamUpdates:
+	default:
+		sm.logger.Error(
+			fmt.Sprintf(
+				"Streaming subscription id %+v channel full capacity. Dropping subscription connection.",
+				subscriptionId,
+			),
+		)
+		removeSubscription = true
 	}
 
-	// Clean up subscriptions that have been closed.
-	// If a Send update has failed for any clob pair id, the whole subscription will be removed.
 	if removeSubscription {
 		sm.removeSubscription(subscriptionId)
 	}
 }
 
+// SendCombinedSnapshot sends messages to a particular subscriber without buffering.
+// Note this method requires the lock and assumes that the lock has already been
+// acquired by the caller.
+func (sm *FullNodeStreamingManagerImpl) SendCombinedSnapshot(
+	offchainUpdates *clobtypes.OffchainUpdates,
+	saUpdates []*satypes.StreamSubaccountUpdate,
+	subscriptionId uint32,
+	blockHeight uint32,
+	execMode sdk.ExecMode,
+) {
+	defer metrics.ModuleMeasureSince(
+		metrics.FullNodeGrpc,
+		metrics.GrpcSendOrderbookSnapshotLatency,
+		time.Now(),
+	)
+
+	var streamUpdates []clobtypes.StreamUpdate
+	streamUpdates = append(streamUpdates, sm.getOrderbookStreamUpdate(offchainUpdates, blockHeight, execMode)...)
+	streamUpdates = append(streamUpdates, sm.getSubaccountStreamUpdates(saUpdates, blockHeight, execMode)...)
```

Comment:
meganit: `getSubaccountStreamUpdate` seems like some getter function for the full node streaming manager, can we rename that to something else or just inline define the struct definitions in this function?

Reply:
renamed to `toSubaccountStreamUpdates` and `toOrderbookStreamUpdate`